src/background.ts

```diff
@@ -70,7 +70,15 @@ class MessageHandler {
 	 */
 	async listAvailablePorts({ origin, options }: BackgroundRequest) {
 		const originAuth = await readOriginAuth(origin)
-		const ports = await listPortsNative()
+		const ports = (await listPortsNative()).filter(port => {
+			return !options || !options.filters || options.filters.filter(filter => {
+				return (!filter.usbVendorId || filter.usbVendorId === port.usb?.vid) &&
+					(!filter.usbProductId || filter.usbProductId === port.usb?.pid) &&
+					(!("id" in filter) || filter["id"] === port.id) &&
+					(!("name" in filter) || filter["name"] === port.name) &&
+					(!("transport" in filter) || filter["transport"] === port.transport)
```

Comment:
Since this is TypeScript, a new type for "extended" filters should be added. The `SerialPortFilter` type can be extended in `types.ts` and each filter can then be typecast to that extended type.

Also, please consider creating another (sub)function like `checkPortMatch(port: SerialPortData, filter: SerialPortFilter): boolean`. Each port could then be checked against each provided filter. The current form is complex to understand, because it uses `!` and `||` with negative logic values.
A function built with `if (filter.<value> && filter.<value> != port.<value>) return false;` would be also less error-prone.

See:
https://wicg.github.io/serial/#dom-serialportfilter